File: backend/app/schemas/arena.py
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field
from app.scenarios.catalog import ScenarioType


class SimulationStatus(StrEnum):
    STOPPED = "stopped"
    RUNNING = "running"
    PAUSED = "paused"


class MarketRegime(StrEnum):
    CALM = "calm"
    VOLATILE = "volatile"
    THIN_LIQUIDITY = "thin_liquidity"


class RedTeamGoal(StrEnum):
    OBVIOUS = "obvious"
    STEALTH = "stealth"
    HARD_TO_DETECT = "hard_to_detect"


class RedTeamScenarioGenerateRequest(BaseModel):
    scenario_family: ScenarioType
    market_regime: MarketRegime
    goal: RedTeamGoal
    constraints: dict[str, Any] = Field(default_factory=dict)


class ScenarioConfig(BaseModel):
    label: str
    slug: ScenarioType
    scenario_family: str
    agent_id: str
    description: str
    launch_endpoint: str
    parameters: dict[str, Any] = Field(default_factory=dict)
    market_regime: MarketRegime
    goal: RedTeamGoal
    source: Literal["nebius", "mock"]
    expected_detector_risk: float = Field(ge=0.0, le=1.0)
    safety_note: str


class BenchmarkResult(BaseModel):
    scenario: str
    precision: float
    recall: float
    f1: float
    avg_detection_latency_ms: float | None = None


class PriceLevel(BaseModel):
    price: float
    quantity: float
    agent_id: str | None = None
    owner: str | None = None
    scenario_id: str | None = None
    scenario_name: str | None = None


class OrderBookSnapshot(BaseModel):
    bids: list[PriceLevel]
    asks: list[PriceLevel]
    best_bid: float | None
    best_ask: float | None
    mid: float | None
    spread: float | None


class MarketFeatures(BaseModel):
    spread_bps: float
    depth_top_n: float
    imbalance: float
    message_rate: float
    cancel_to_trade_ratio: float
    order_lifetime_ms: float
    wall_size_ratio: float
    depth_change_pct: float
    distance_from_touch_bps: float = 0.0
    cancel_probability: float = 0.0
    execution_ratio: float = 0.0
    replenishment_rate: float = 0.0
    side_switching_rate: float = 0.0
    participant_order_linkage: float = 0.0
    linked_participant_ids: str = ""
    linked_order_ids: str = ""
    linked_event_ids: str = ""
    large_level_count: float = 0.0


class DetectorScore(BaseModel):
    name: str
    confidence: float = Field(ge=0.0, le=1.0)
    alert: bool
    severity: Literal["low", "medium", "high", "critical"] | None = None
    evidence: list["EvidenceItem"] | None = None


class DetectorScores(BaseModel):
    scores: list[DetectorScore]
    alerts: list[DetectorScore]


class AgentEvent(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: str
    event_id: str | None = None
    timestamp: float | None = None
    order_id: str | None = None
    aggressor_order_id: str | None = None
    resting_order_id: str | None = None
    agent_id: str | None = None
    aggressor_agent_id: str | None = None
    resting_agent_id: str | None = None
    side: Literal["buy", "sell"] | None = None
    price: float | None = None
    quantity: float | None = None
    scenario_id: str | None = None
    scenario_name: str | None = None
    scenario_family: str | None = None


class EvidenceItem(BaseModel):
    key: str
    label: str
    value: str | int | float | bool
    unit: str | None = None
    interpretation: str | None = None


class Incident(BaseModel):
    id: str
    title: str
    type: str
    agent: str
    confidence: float = Field(ge=0.0, le=1.0)
    severity: Literal["Low", "Medium", "High", "Critical"]
    evidence: list[EvidenceItem]
    explanation: str
    scenario_id: str | None = None
    scenario_family: str | None = None


class AttackStage(StrEnum):
    ARMED = "armed"
    WALL_PLACED = "wall_placed"
    PRESSURE_PHASE = "pressure_phase"
    WALL_CANCELLED = "wall_cancelled"
    CANCELLED = "cancelled"
    INCIDENT_CONFIRMED = "incident_confirmed"
    DONE = "done"


class AttackStageSnapshot(BaseModel):
    detector_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    label: str
    stage: AttackStage
    status: Literal["pending", "active", "completed"]
    tick: int | None = None
    timestamp: float | None = None


class ScenarioLabel(BaseModel):
    label_id: str
    run_id: str
    scenario_id: str
    scenario_family: str
    scenario_name: str
    seed: int
    start_tick: int
    expected_end_tick: int | None = None
    actual_end_tick: int | None = None
    agent_ids: list[str]
    event_ids: list[str] = Field(default_factory=list)
    order_ids: list[str] = Field(default_factory=list)
    manipulation_windows: list[dict[str, int | None]] = Field(default_factory=list)
    phase_windows: dict[str, dict[str, int | None]] = Field(default_factory=dict)
    parameters: dict[str, Any] = Field(default_factory=dict)


class AttackTrackerState(BaseModel):
    scenario_id: str
    scenario_name: str
    scenario_family: str
    agent_id: str
    current_stage: AttackStage | None = None
    start_tick: int
    stages: list[AttackStageSnapshot] | None = None
    status: AttackStage | str
    evidence: list[EvidenceItem] | None = None
    label: ScenarioLabel | None = None


class ExchangeEventRecord(BaseModel):
    schema_version: int = 1
    event_type: Literal["add", "modify", "cancel", "execute", "snapshot"]
    event_id: str
    sequence: int = Field(ge=1)
    source: Literal["simulation", "historical"]
    source_sequence: int | None = None
    symbol: str
    venue: str
    tick: int | None = None
    exchange_timestamp_ns: int | None = None
    received_timestamp_ns: int | None = None
    scenario_id: str | None = None
    scenario_name: str | None = None
    scenario_family: str | None = None
    order_id: str | None = None
    agent_id: str | None = None
    side: Literal["buy", "sell"] | None = None
    price: float | None = None
    quantity: float | None = None
    owner: str | None = None
    previous_price: float | None = None
    previous_quantity: float | None = None
    priority_preserved: bool | None = None
    execution_id: str | None = None
    aggressor_order_id: str | None = None
    resting_order_id: str | None = None
    aggressor_agent_id: str | None = None
    resting_agent_id: str | None = None
    aggressor_remaining_quantity: float | None = None
    resting_remaining_quantity: float | None = None
    depth: int | None = None
    book: OrderBookSnapshot | None = None


class ExchangeEventReplay(BaseModel):
    events: list[ExchangeEventRecord]
    after_sequence: int = Field(ge=0)
    next_after_sequence: int = Field(ge=0)
    latest_sequence: int = Field(ge=0)
    stream_id: str | None = None
    first_available_sequence: int | None = Field(default=None, ge=0)
    retained_from_sequence: int | None = Field(default=None, ge=0)
    has_more: bool


class ArenaMetricsSnapshot(BaseModel):
    tick: int = Field(ge=0)
    running: bool
    incidents_count: int = Field(ge=0)


class ArenaState(BaseModel):
    tick: int
    running: bool
    events: list[AgentEvent]
    exchange_events: list[ExchangeEventRecord] = Field(default_factory=list)
    book: OrderBookSnapshot
    best_bid: float | None
    best_ask: float | None
    mid: float | None
    spread: float | None
    active_agents: list[str]
    active_scenario: AttackTrackerState | None
    detectors: DetectorScores
    incidents: list[Incident] | None = None
    features: dict[str, Any] | MarketFeatures | None = None
    historical_events: list[dict[str, Any]] = Field(default_factory=list)
    market_data: dict[str, Any] | None = None

File: frontend/src/types/arena.ts
export type PriceLevel = {
  agent_id?: string;
  owner?: "normal" | "abuser" | string;
  price: number;
  quantity: number;
  scenario_id?: string;
  scenario_name?: string;
};

export type OrderBookSnapshot = {
  bids: PriceLevel[];
  asks: PriceLevel[];
  best_bid: number | null;
  best_ask: number | null;
  mid: number | null;
  spread: number | null;
};

export type MarketFeatures = {
  spread_bps: number;
  depth_top_n: number;
  imbalance: number;
  order_book_imbalance?: number;
  top_n_bid_depth?: number;
  top_n_ask_depth?: number;
  message_rate: number;
  message_rate_per_sec?: number;
  cancel_to_trade_ratio: number;
  order_lifetime_ms: number;
  wall_size_ratio: number;
  large_level_count?: number;
  depth_change_pct: number;
};

export type DetectorScore = {
  name: string;
  confidence: number;
  alert: boolean;
  severity?: "low" | "medium" | "high" | "critical";
  evidence?: EvidenceItem[];
};

export type DetectorScores = {
  scores: DetectorScore[];
  alerts: DetectorScore[];
};

export type AgentEvent = {
  type: string;
  tick?: number;
  timestamp?: number;
  order_id?: string;
  aggressor_order_id?: string;
  resting_order_id?: string;
  agent_id?: string;
  aggressor_agent_id?: string;
  resting_agent_id?: string;
  side?: "buy" | "sell";
  price?: number;
  quantity?: number;
  scenario_id?: string | null;
  scenario_name?: string | null;
  scenario_family?: string | null;
  runtime_source?: "backend" | "agent_runner";
  [key: string]: unknown;
};

export type ExchangeEventType = "add" | "modify" | "cancel" | "execute" | "snapshot";
export type ExchangeEventSource = "simulation" | "historical";

type ExchangeEventBase = {
  schema_version: 1;
  event_type: ExchangeEventType;
  event_id: string;
  sequence: number;
  source: ExchangeEventSource;
  source_sequence: number | null;
  symbol: string;
  venue: string;
  tick: number | null;
  exchange_timestamp_ns: number | null;
  received_timestamp_ns: number | null;
  scenario_id: string | null;
  scenario_name: string | null;
  scenario_family: string | null;
};

type RestingOrderPayload = {
  order_id: string;
  agent_id: string;
  side: "buy" | "sell";
  price: number;
  quantity: number;
  owner: string;
};

export type AddExchangeEvent = ExchangeEventBase & RestingOrderPayload & {
  event_type: "add";
};

export type ModifyExchangeEvent = ExchangeEventBase & RestingOrderPayload & {
  event_type: "modify";
  previous_price: number;
  previous_quantity: number;
  priority_preserved: boolean;
};

export type CancelExchangeEvent = ExchangeEventBase & RestingOrderPayload & {
  event_type: "cancel";
};

export type ExecuteExchangeEvent = ExchangeEventBase & {
  event_type: "execute";
  execution_id: string;
  aggressor_order_id: string;
  resting_order_id: string;
  aggressor_agent_id: string;
  resting_agent_id: string;
  side: "buy" | "sell";
  price: number;
  quantity: number;
  aggressor_remaining_quantity: number;
  resting_remaining_quantity: number;
};

export type SnapshotExchangeEvent = ExchangeEventBase & {
  event_type: "snapshot";
  depth: number;
  book: OrderBookSnapshot;
};

export type ExchangeEvent =
  | AddExchangeEvent
  | ModifyExchangeEvent
  | CancelExchangeEvent
  | ExecuteExchangeEvent
  | SnapshotExchangeEvent;

export type EvidenceItem = {
  key: string;
  label: string;
  value: string | number | boolean;
  unit?: string;
  interpretation?: string;
};

export type InvestigationContext = {
  simulation_metadata?: Record<string, unknown>;
  market_regime?: Record<string, unknown>;
  instrument?: Record<string, unknown>;
  episode_duration?: Record<string, unknown>;
  suspected_agent?: Record<string, unknown>;
  order_book_context?: Record<string, unknown>;
  trades?: Record<string, unknown>[];
  event_timeline?: Record<string, unknown>[];
  market_metrics?: Record<string, unknown>;
  cancellation_metrics?: Record<string, unknown>;
  execution_metrics?: Record<string, unknown>;
  price_movement?: Record<string, unknown>;
  ground_truth?: Record<string, unknown> | null;
  previous_agent_behaviour?: Record<string, unknown>[];
};

export type Incident = {
  id: string;
  title: string;
  type: string;
  agent: string;
  confidence: number;
  severity: "Low" | "Medium" | "High" | "Critical";
  evidence: EvidenceItem[];
  explanation: string;
  scenario_id?: string;
  scenario_family?: string;
  tick?: number;
  investigation_context?: InvestigationContext;
};

export type AttackStage =
  | "armed"
  | "wall_placed"
  | "pressure_phase"
  | "wall_cancelled"
  | "cancelled"
  | "incident_confirmed"
  | "done";

export type AttackStageStatus = "pending" | "active" | "completed";

export type AttackStageSnapshot = {
  detector_confidence?: number;
  label: string;
  stage: AttackStage;
  status: AttackStageStatus;
  tick?: number;
  timestamp?: number;
};

export type AttackTrackerState = {
  scenario_id: string;
  scenario_name: string;
  scenario_family: string;
  agent_id: string;
  current_stage?: AttackStage;
  start_tick: number;
  stages?: AttackStageSnapshot[];
  status: AttackStage | string;
  evidence?: EvidenceItem[];
};

export type ArenaState = {
  tick: number;
  running: boolean;
  events: AgentEvent[];
  exchange_events: ExchangeEvent[];
  book: OrderBookSnapshot;
  best_bid: number | null;
  best_ask: number | null;
  mid: number | null;
  spread: number | null;
  active_agents: string[];
  active_scenario: AttackTrackerState | null;
  detectors: DetectorScores;
  incidents?: Incident[];
  features?: Partial<MarketFeatures>;
  historical_events?: Array<HistoricalMarketEvent | ExchangeEvent>;
  market_data?: MarketDataContext;
};

export type HistoricalMarketEvent = {
  type: string;
  event_kind: string;
  source_event_code: number;
  source_sequence: number;
  timestamp_ns_since_midnight: number;
  order_id: string;
  source_order_id: number;
  quantity: number;
  price_x10000: number;
  price?: number;
  side?: "buy" | "sell" | null;
  symbol: string;
  source: "historical";
};

export type MarketDataContext = {
  source_type: "historical" | "hybrid" | "synthetic_profile";
  dataset_id: string;
  format?: string;
  symbol: string;
  venue?: string;
  trade_date: string;
  depth: number;
  source_sequence: number;
  replay_position: number;
  exchange_timestamp_ns: number;
  row_count: number;
  progress: number;
  eof: boolean;
  events_sha256?: string;
  profile_id?: string;
  profile_sha256?: string;
  training_dataset_id?: string;
  current_reference_price_ticks?: number;
  run_binding_sha256?: string;
};

export type ArenaWebSocketMessage = {
  type: "arena_state";
  version: number;
  timestamp: string;
  payload: ArenaState;
};

export type ScenarioConfig = {
  label: string;
  slug: string;
  scenario_family: string;
  agent_id: string;
  description: string;
  launch_endpoint?: string;
  parameters?: Record<string, unknown>;
  market_regime?: "calm" | "volatile" | "thin_liquidity";
  goal?: "obvious" | "stealth" | "hard_to_detect";
  source?: "nebius" | "mock";
  expected_detector_risk?: number;
  safety_note?: string;
};

export type BenchmarkResult = {
  avg_detection_latency_ms?: number;
  scenario: string;
  precision: number;
  recall: number;
  f1: number;
};
